src/sse/sse_utils.rs

```diff
@@ -0,0 +1,265 @@
+            }
+        };
+        Self {
+            //sign_lo,
+            sign_hi,
+            sign_both,
+        }
+    }
+
+    #[inline(always)]
+    pub unsafe fn rotate_hi(&self, values: __m128) -> __m128 {
+        let temp = _mm_shuffle_ps(values, values, 0xB4);
+        _mm_xor_ps(temp, self.sign_hi)
+    }
+
+    // There doesn't seem to be any need for rotating just the first element, but let's keep the code just in case
+    //#[inline(always)]
+    //pub unsafe fn rotate_lo(&self, values: __m128) -> __m128 {
+    //    let temp = _mm_shuffle_ps(values, values, 0xE1);
+    //    _mm_xor_ps(temp, self.sign_lo)
+    //}
+
+    #[inline(always)]
+    pub unsafe fn rotate_both(&self, values: __m128) -> __m128 {
+        let temp = _mm_shuffle_ps(values, values, 0xB1);
+        _mm_xor_ps(temp, self.sign_both)
+    }
+}
+
+// Pack low (1st) complex
+// left: r1.re, r1.im, r2.re, r2.im
+// right: l1.re, l1.im, l2.re, l2.im
+// --> r1.re, r1.im, l1.re + l1.im
+#[inline(always)]
+pub unsafe fn extract_lo_lo_f32(left: __m128, right: __m128) -> __m128 {
+    //_mm_shuffle_ps(left, right, 0x44)
+    _mm_castpd_ps(_mm_unpacklo_pd(_mm_castps_pd(left), _mm_castps_pd(right)))
+}
+
+// Pack high (2nd) complex
+// left: r1.re, r1.im, r2.re, r2.im
+// right: l1.re, l1.im, l2.re, l2.im
+// --> r2.re, r2.im, l2.re + l2.im
+#[inline(always)]
+pub unsafe fn extract_hi_hi_f32(left: __m128, right: __m128) -> __m128 {
+    //_mm_shuffle_ps(left, right, 0xEE)
+    _mm_castpd_ps(_mm_unpackhi_pd(_mm_castps_pd(left), _mm_castps_pd(right)))
+}
+
+// Pack low (1st) and high (2nd) complex
+// left: r1.re, r1.im, r2.re, r2.im
+// right: l1.re, l1.im, l2.re, l2.im
+// --> r1.re, r1.im, l2.re + l2.im
+#[inline(always)]
+pub unsafe fn extract_lo_hi_f32(left: __m128, right: __m128) -> __m128 {
+    //_mm_shuffle_ps(left, right, 0xE4)
+    _mm_blend_ps(left, right, 0x0C)
+}
+
+// Pack  high (2nd) and low (1st) complex
+// left: r1.re, r1.im, r2.re, r2.im
+// right: l1.re, l1.im, l2.re, l2.im
+// --> r2.re, r2.im, l1.re + l1.im
+#[inline(always)]
+pub unsafe fn extract_hi_lo_f32(left: __m128, right: __m128) -> __m128 {
+    _mm_shuffle_ps(left, right, 0x4E)
+}
+
+// Reverse complex
+// values: a.re, a.im, b.re, b.im
+// --> b.re, b.im, a.re + a.im
+#[inline(always)]
+pub unsafe fn reverse_complex_elements_f32(values: __m128) -> __m128 {
+    _mm_shuffle_ps(values, values, 0x4E)
+}
+
+// Invert sign of high (2nd) complex
+// values: a.re, a.im, b.re, b.im
+// -->  a.re, a.im, -b.re, -b.im
+#[inline(always)]
+pub unsafe fn negate_hi_f32(values: __m128) -> __m128 {
+    _mm_xor_ps(values, _mm_set_ps(-0.0, -0.0, 0.0, 0.0))
+}
+
+// Duplicate low (1st) complex
+// values: a.re, a.im, b.re, b.im
+// --> a.re, a.im, a.re + a.im
+#[inline(always)]
+pub unsafe fn duplicate_lo_f32(values: __m128) -> __m128 {
+    _mm_shuffle_ps(values, values, 0x44)
+}
+
+// Duplicate high (2nd) complex
+// values: a.re, a.im, b.re, b.im
+// --> b.re, b.im, b.re + b.im
```

Comment:
Addition sign

Reply:
Fixed, and I found several more. I think now they are finally all gone.